### PR TITLE
INT: Add namespace intention #489 for trivial cases

### PR DESCRIPTION
**Pull request by la10736**
I implemented the intention just in the trivial but frequent case where you have just one generic parameter. I wrote a test case for the general case but comment it out. I don't know how can I try to solve it... any hints will be very appreciated. Maybe it is better spread general case in another PR.

### CODE REVIEW

**Comment by matklad** [comment]
Nice! I think that instead of the intention, this should be a quick fix associated with the error annotation.

You can add a quick fix like this:

```Kotlin
    private fun checkBinary(holder: AnnotationHolder, o: RsBinaryExpr) {
        if (o.isComparisonBinaryExpr() && (o.left.isComparisonBinaryExpr() || o.right.isComparisonBinaryExpr())) {
            holder.createErrorAnnotation(o, "Chained comparison operator require parentheses")
                .registerFix(AddTurbofishOperatorFix())
        }
    }
```

**Comment by matklad** [comment]
Oups, didn't mean to close this :sweat_smile:

File: src/main/kotlin/org/rust/ide/intentions/AddNamespaceIntention.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.util.parentOfType
+import org.rust.lang.core.psi.RsElementTypes.*
+
+class AddNamespaceIntention : RsElementBaseIntentionAction<AddNamespaceIntention.Context>() {
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val (matchExpr, caller, gen, more) = ctx
+        val callWithNamespace =
+            RsPsiFactory(project).createExpression("""${caller.text}::<${gen.text}>${more.text}""") as RsCallExpr
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/intentions/AddNamespaceIntention.kt]: Single quotes should be enough here. Also you don't need this cast

File: src/main/kotlin/org/rust/ide/intentions/AddNamespaceIntention.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    )
+
+    override fun getText() = "Add missed namespace"
+    override fun getFamilyName() = text
+
+    private fun resolveMatchExpression(element: PsiElement): RsBinaryExpr? {
+        val base = element.parentOfType<RsBinaryExpr>() ?: return null
+        return if (base.left is RsBinaryExpr) {
+                base
+            } else {
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/intentions/AddNamespaceIntention.kt]: Indentation is a bit messed up. Reformat file (`Ctrl+Alt+l`) should cleanup this.

**Comment by matklad** [comment]
> I don't know how can I try to solve it... any hints will be very appreciated.

That's an interesting problem indeed!  When there are several type arguments, you don't get any reasonable PSI tree because the parser expects expression after the `<` and `T, U` does not look like an expression at all:

train_wreck

The good news is that we still trigger "chain comparison" annotator in this cases sometimes. For example, for 

```
.collect<Result<Vec<_>, _>>();
```

We will parse `< Result <` part. I think we can handle this, if we base our modifications not on the `PSI`, but on the text instead. That is, 

```
editor.document.insertString(lt.textOffset - 1, "::")
```

should work in more cases. 


This still won't handle the case like `.collect<T, U>();`, but the problem here is that even the original error annotation does not work here, because we fail to reconstruct any reasonable PSI. I suggest we don't handle this case in this PR.

**Comment by matklad** [comment]
>but the problem here is that even the original error annotation does not work here, because we fail to reconstruct any reasonable PSI. 

How should we fix it? I don't know a good solution unfortunately. The nice trick would be to check all `a < b` expressions, where the node immediately after `b` is an error element. Then, we can try to copy the file, insert `::` before `<` and see if that fixes syntax errors. If it does, we can suggest a fix.

**Comment by la10736** [comment]
@matklad 
> You can add a quick fix like this:

I didn't know as do it... Is it possible also filter the cases where register the quick fix? I'm not sure if it is a good idea try to fix all chained comparison.

**Comment by matklad** [comment]
>I didn't know as do it... Is it possible also filter the cases where register the quick fix?

Sure! The code should like this:

```
val annotation = holder.createAnnotation()

if (looksLikeTheUserForgotTheTurbofish) {
    annotation.registerFix(RsAddTurbofishOperatorFix())
}
```

**Comment by la10736** [comment]
I don't think we should relay on the *lucky* case where chained comparison is triggered. Is better try to write something that is quite closer this issue like a  `,` in a `PsiErrorElement` and find a `RsBinaryExpression` thats contain it in the right branch. I think that can be more robust.

I'll try to handle this special cases in an other PR. I'll take in account your considerations and fix this simple case by use quickfix instead annotation (maybe is better a new PR too to take commits clean).

**Comment by matklad** [comment]
>I don't think we should relay on the lucky case where chained comparison is triggered.

My though was that we can fix the "chained comparison" to trigger in more cases. If we can fix something, we must be aware that the problem exists anyway :) 

>maybe is better a new PR too to take commits clean

That sounds reasonable! Or you can just force-push to this branch.

**Comment by la10736** [comment]
I'm working (slowly) on handling some more cases. I'll move these changes on another PR.

**Pull request merged**